Add support for defining included column for indexes in SQL server

Kukkimonsuta opened the pull request:
This PR adds `ForSqlServerInclude` extension to `IndexBuilder` allowing defining columns to be included in a index on sql server.

To support expression parameter a generic version of `IndexBuilder` is introduced in main EF assembly, however to avoid breaking changes (as requested in https://github.com/aspnet/EntityFrameworkCore/issues/4846#issuecomment-395213518) this type is available only with new extension `ForSqlServerHasIndex` for the time being.

There is a breaking change in `EFCore.SqlServer` adding a property to interface -
 `ISqlServerIndexAnnotations.Include`, which might be ok since it's maybe not really meant to be implemented by others? @ajcvickers 

I'm still investigating possible problems with migrations, so this is not yet ready.

Fixes #4846

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/aspnet/EntityFrameworkCore?pullRequest=12266) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/aspnet/EntityFrameworkCore?pullRequest=12266) before we can accept your contribution.<br/><br/>:x: Kukkimonsuta [sign now](https://cla.dotnetfoundation.org/aspnet/EntityFrameworkCore?pullRequest=12266)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/aspnet/EntityFrameworkCore?pullRequest=12266) it.</sub>

Kukkimonsuta commented:
Rebased onto latest dev.

One thing I think is worth noting is that `property.GetContainingIndexes` will not return indexes that have this property in include clause. I think it makes sense and sql generator 
works around this by scanning whole entity instead ( https://github.com/aspnet/EntityFrameworkCore/pull/12266/commits/f236b8c43e3d7896d2d4f7bd2031040dbfbc7746#diff-42efb7d65f79f8f8347d8753fcee5498R1683 ), but it's certainly possible it may cause trouble elsewhere.

Other than that I have not found any further issues, so this should be ready for review.

AndriySvyryd commented on src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs:
```diff
@@ -1896,7 +1896,7 @@ protected virtual bool HasDifferences([NotNull] IEnumerable<IAnnotation> source,
             foreach (var annotation in source)
             {
-                var index = unmatched.FindIndex(a => a.Name == annotation.Name && Equals(a.Value, annotation.Value));
+                var index = unmatched.FindIndex(a => a.Name == annotation.Name && AnnotationValueEquals(a.Value, annotation.Value));
```
Just call `StructuralComparisons.StructuralEqualityComparer.Equals`

Kukkimonsuta commented:
1. added generic variants of `HasAnnotation`, `IsUnique`, `HasName`, `HasFilter`, `ForSqlServerIsClustered` to index builder
2. changed differ to use structural comparison for annotations
3. renamed `Include` to `IncludeProperties` (I've used `includeProperties` everywhere for consitency)
4. add validation of duplicate/missing/already included properties

Kukkimonsuta commented:
Rebased and squashed.

AndriySvyryd commented:
@Kukkimonsuta Thanks for your contribution!
